### ISSUE TITLE
File name mismatch in Tika/Freedesktop.org ids

### ISSUE THREAD

**Issue by ross-spencer** [richardlehane/siegfried#78]
Hi Richard,

Still working away at my analysis scripts for SF. (Getting close!) 

I'm trying to understand basis today for the non PRONOM IDS. Is filename mistmatch the same as extension mismatch in DROID or is there a subtlety I'm missing? 

It seems to me that it's simply reporting this for when there is no extension? 

       filename : '/home/goatslayer/git/droid-sqlite-analysis/.git/description'
        - ns      : tika
          id      : 'text/plain'
          format  : 
          mime    : 'text/plain'
          basis   : 'text match ASCII'
          warning : 'match on text only; byte/xml signatures for this format did not match; filename mismatch'
        - ns      : freedesktop.org
          id      : 'text/plain'
          format  : 'plain text document'
          mime    : 'text/plain'
          basis   : 'text match ASCII'
          warning : 'match on text only; byte/xml signatures for this format did not match; filename mismatch'

**Comment by richardlehane** [comment]
Hey Ross, 
they're *almost* the same.

The MIMEInfo spec doesn't have extension signatures, it has filename [glob](https://en.wikipedia.org/wiki/Glob_(programming)) signatures. The majority of these are really just extensions (e.g. "*.txt") but some aren't e.g. ("README", "COPYING", etc. can also be signatures). 

From 1.5, the old extension matcher was replaced with a new namematcher matcher that can do glob matching as well as extension matching. The basis now reports whether it is an extension or a glob match ([see here](https://github.com/richardlehane/siegfried/blob/master/pkg/core/namematcher/namematcher.go#L216)). For the purpose of mismatch reporting, I don't know if it is a glob or extension match that should have happened, which is why you get "filename" mismatch to cover both possibilities.

**Comment by ross-spencer** [comment]
Neat feature!

I've managed to grab an example with a Glob match for testing. The idea of
a filename mismatch makes sense. Will figure out how best to report on
these.

Thanks Richard.

**Comment by ross-spencer** [comment]
For an ID like this I'm wondering if the Warning wouldn't be better reported as 'extension only' as with the PRONOM ids, and the basis:

      ---
      filename : '/home/goatslayer/git/droid-sqlite-analysis/textoutputclass.pyc'
      filesize : 7700
      modified : 2016-03-17T00:47:42+13:00
      errors   : 
      md5      : e8d220b71eab5723f82dce55c7120b9a
      matches  :
        - ns      : pronom
          id      : UNKNOWN
          format  : 
          version : 
          mime    : 
          basis   : 
          warning : 'no match'
        - ns      : tika
          id      : 'UNKNOWN'
          format  : 
          mime    : 'UNKNOWN'
          basis   : 
          warning : 'no match'
        - ns      : freedesktop.org
          id      : 'application/x-python-bytecode'
          format  : 'Python bytecode'
          mime    : 'application/x-python-bytecode'
          basis   : 'extension match pyc'
          warning : 'match on filename only; byte/xml signatures for this format did not match'

**Comment by richardlehane** [comment]
Hi Ross,
I'm not sure I follow ... are you suggesting changning the freedesktop warning ('match on filename only; byte/xml signatures for this format did not match')? I.e. to get rid of the latter bit?

The reason for the text 'byte/xml signatures for this format did not match' is to warn the user that that format has a possible byte signature (value="0x994e0d0a" type="big32" offset="0") that hasn't matched for your particular file.

This behaviour is an example of where the MIMEInfo identifier operates differently from the PRONOM identifier. For the PRONOM identifier, the failure of a byte signature to match, like happened there, disqualifies format match even if there was a possible extension match because the absence of those bytes should mean it definitely isn't that format (or something is wrong with the file).

The MIMEInfo identifier initially worked in the same way (so you would have got an unknown there instead of the python bytecode match). But I found in practice that the MIMEInfo byte and xml signatures often are just optional and the absence of those signatures didn't necessarily mean the files weren't of the type indicated by their extension. E.g., in your example, you potentially can have a python bytecode file that doesn't have that magic 0x994e0d0a at the start.

Does that make sense??

**Comment by ross-spencer** [comment]
That's cool... my query was more about the first part, and I might just be trying to understand the difference in terminology compared to what i'm used to in asking but in PRONOM the following would happen for a blank text file: 

      filename : 'blank.txt'
      filesize : 0
      modified : 2016-04-05T10:38:01+12:00
      errors   : 'empty source'
      matches  :
        - ns      : pronom
          id      : x-fmt/111
          format  : 'Plain Text File'
          version :
          mime    : 'text/plain'
          basis   : 'extension match txt'
          warning : 'match on extension only'

So our basis is extension match, and the warning is extension only. 

In the freedesktop example for the Python Byte Code we have: 

    - ns      : freedesktop.org
        id      : 'application/x-python-bytecode'
        format  : 'Python bytecode'
        mime    : 'application/x-python-bytecode'
        basis   : 'extension match pyc'
        warning : 'match on filename only; byte/xml signatures for this format did not match'

Basis is extension match, but warning is filename only. 

I can see this probably relates to what you were saying about GLOB and filename matching. i was just wondering if the warning would benefit from being more consistent/specific if that were possible? - and so we know it's not a filename match even as bytecode files wouldn't be consistently named to have filename matches, but we do know it's only an extension match. 

IDK! ^_^

**Comment by richardlehane** [comment]
Ahh, I was worried I might be going off on a tangent!

It might be possible to be more specific with the warning but it may also be a little tricky!

Both the PRONOM and MIMEInfo identifiers both have their own "Identification" structs:

  - MIMEInfo: https://github.com/richardlehane/siegfried/blob/master/pkg/mimeinfo/identifier.go#L541
  - PRONOM: https://github.com/richardlehane/siegfried/blob/master/pkg/pronom/identifier.go#L441

When matches come in from the matching routines, we just add these Identifications to a slice. If an Identification is added twice (e.g. because both extension and byte sigs match), then the score for that particular Identification is increased. In PRONOM we store that in the "confidence" int. The MIMEInfo score is a composite score made up of a number of fields (xmlMatch, magicScore, globScore etc.). At the end, those slices are then sorted (for PRONOM that is easy - just sort on that int, the sorting algo for MIMEInfo looks at all those different bits of the score) and we just take the highest scoring ones for reporting back to the user.

The basis for the matches is stored in the Basis string slices for both Identifications. These strings come directly from the matchers (which is why they know such detail about matches e.g the offsets where byte matches occur).

The warnings aren't generated in relation to those Basis strings at all (I would have to do all this string parsing that you are currently doing, that is horrible! ha ha) - the warnings come from the level of "confidence". In the PRONOM code, we just infer from the confidence score itself what that warning should be (the lowConfidence func here just takes that int as input and returns a string: https://github.com/richardlehane/siegfried/blob/master/pkg/pronom/identifier.go#L316). The MIMEInfo code (https://github.com/richardlehane/siegfried/blob/master/pkg/mimeinfo/identifier.go#L502) is a little more convoluted but works similarly - it looks at all those confidence signals to work out what an appropriate warning should be. Unfortunately the only signal we have for the namematcher is that globscore field & this doesn't differentiate between ext/glob. So in order to be more specific in the warning we'd need to either start parsing the Basis slice (yuck) or store more information in the Identification struct (also a bit yuck).

I.e. it is a good suggestion but may be a bit of a hack to change... I'll look at it a bit further for next release!

**Comment by ross-spencer** [comment]
Thanks Richard. I appreciate that. My instinct is that a little of how we start to interpret your results is perhaps more about understanding and education than something that needs fixing - and so I guess I'm just beginning on that pathway now. if it is something you feel might be beneficial then I'd be interested to see the change, but that being said I'm just happy to work with any of the data coming my way. 

As for my Py code then yes - horrible is the word! I've described it as a cat trying to squeeze into a vase - it's possible but it's not always pretty. I'm close to getting it all into to DB though and once I've got that baseline I'll refine... I'm just understanding all the data at the moment - it's interesting! I think we'll have something useful by the beginning of next week. I hope you'll like it! - I'm learning a lot. A bit of a sidetrack there. 

Thanks for the extra information and links to Go snippets. All useful.
